MediaPlayerPrivateWebM doesn't handle decoding errors.

jyavenard opened the pull request:
#### ac3459ad6236a13dcf9c63728c820c6eb722f56d
<pre>
MediaPlayerPrivateWebM doesn&apos;t handle decoding errors.
<a href="https://bugs.webkit.org/show_bug.cgi?id=267165">https://bugs.webkit.org/show_bug.cgi?id=267165</a>
<a href="https://rdar.apple.com/120570295">rdar://120570295</a>

Reviewed by Youenn Fablet.

The MediaPlayerPrivateWebM didn&apos;t listen for error from either the AVSampleBufferDisplayLayer
nor the AVSampleBufferAudioRenderer.
We extract the code from the SourceBufferPrivateAVFObjC and create a new WebAVSampleBufferListener
class.

Add test.

* LayoutTests/media/content/garbage.webm: Added.
* LayoutTests/media/media-vp8-webm-error-expected.txt: Added.
* LayoutTests/media/media-vp8-webm-error.html: Added.
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/avfoundation/WebAVSampleBufferListener.h: Added.
(WebCore::WebAVSampleBufferListenerClient::layerDidReceiveError):
(WebCore::WebAVSampleBufferListenerClient::layerRequiresFlushToResumeDecodingChanged):
(WebCore::WebAVSampleBufferListenerClient::layerReadyForDisplayChanged):
(WebCore::WebAVSampleBufferListenerClient::rendererDidReceiveError):
(WebCore::WebAVSampleBufferListenerClient::rendererWasAutomaticallyFlushed):
(WebCore::WebAVSampleBufferListenerClient::outputObscuredDueToInsufficientExternalProtectionChanged):
* Source/WebCore/platform/graphics/avfoundation/WebAVSampleBufferListener.mm: Added.
(-[WebAVSampleBufferListenerPrivate initWithClient:]):
(-[WebAVSampleBufferListenerPrivate invalidate]):
(-[WebAVSampleBufferListenerPrivate beginObservingLayer:]):
(-[WebAVSampleBufferListenerPrivate stopObservingLayer:]):
(-[WebAVSampleBufferListenerPrivate beginObservingRenderer:]):
(-[WebAVSampleBufferListenerPrivate stopObservingRenderer:]):
(-[WebAVSampleBufferListenerPrivate observeValueForKeyPath:ofObject:change:context:]):
(-[WebAVSampleBufferListenerPrivate layerFailedToDecode:]):
(-[WebAVSampleBufferListenerPrivate layerRequiresFlushToResumeDecodingChanged:]):
(-[WebAVSampleBufferListenerPrivate layerReadyForDisplayChanged:]):
(-[WebAVSampleBufferListenerPrivate audioRendererWasAutomaticallyFlushed:]):
(WebCore::WebAVSampleBufferListener::WebAVSampleBufferListener):
(WebCore::WebAVSampleBufferListener::invalidate):
(WebCore::WebAVSampleBufferListener::beginObservingLayer):
(WebCore::WebAVSampleBufferListener::stopObservingLayer):
(WebCore::WebAVSampleBufferListener::beginObservingRenderer):
(WebCore::WebAVSampleBufferListener::stopObservingRenderer):
* Source/WebCore/platform/graphics/avfoundation/objc/AudioTrackPrivateMediaSourceAVFObjC.cpp:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::SourceBufferPrivateAVFObjC):
(WebCore::SourceBufferPrivateAVFObjC::destroyRenderers):
(WebCore::SourceBufferPrivateAVFObjC::trackDidChangeEnabled):
(WebCore::SourceBufferPrivateAVFObjC::setVideoLayer):
(-[WebAVSampleBufferListener initWithParent:]): Deleted.
(-[WebAVSampleBufferListener dealloc]): Deleted.
(-[WebAVSampleBufferListener invalidate]): Deleted.
(-[WebAVSampleBufferListener beginObservingLayer:]): Deleted.
(-[WebAVSampleBufferListener stopObservingLayer:]): Deleted.
(-[WebAVSampleBufferListener beginObservingRenderer:]): Deleted.
(-[WebAVSampleBufferListener stopObservingRenderer:]): Deleted.
(-[WebAVSampleBufferListener observeValueForKeyPath:ofObject:change:context:]): Deleted.
(-[WebAVSampleBufferListener layerFailedToDecode:]): Deleted.
(-[WebAVSampleBufferListener layerRequiresFlushToResumeDecodingChanged:]): Deleted.
(-[WebAVSampleBufferListener layerReadyForDisplayChanged:]): Deleted.
(-[WebAVSampleBufferListener audioRendererWasAutomaticallyFlushed:]): Deleted.
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
(WebCore::MediaPlayerPrivateWebM::getWeakPtr const):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::MediaPlayerPrivateWebM):
(WebCore::MediaPlayerPrivateWebM::~MediaPlayerPrivateWebM):
(WebCore::MediaPlayerPrivateWebM::setDuration):
(WebCore::MediaPlayerPrivateWebM::notifyClientWhenReadyForMoreSamples):
(WebCore::MediaPlayerPrivateWebM::provideMediaData):
(WebCore::MediaPlayerPrivateWebM::trackDidChangeSelected):
(WebCore::MediaPlayerPrivateWebM::didParseInitializationData):
(WebCore::MediaPlayerPrivateWebM::append):
(WebCore::MediaPlayerPrivateWebM::ensureLayer):
(WebCore::MediaPlayerPrivateWebM::ensureDecompressionSession):
(WebCore::MediaPlayerPrivateWebM::addAudioRenderer):
(WebCore::MediaPlayerPrivateWebM::destroyLayer):
(WebCore::MediaPlayerPrivateWebM::destroyAudioRenderer):
(WebCore::MediaPlayerPrivateWebM::registerNotifyWhenHasAvailableVideoFrame):
(WebCore::MediaPlayerPrivateWebM::startVideoFrameMetadataGathering):
(WebCore::MediaPlayerPrivateWebM::layerDidReceiveError):
(WebCore::MediaPlayerPrivateWebM::rendererDidReceiveError):
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::finishCurrentDecodingAndReportError):

should we continue to decode data, a new error won&apos;t be produced.

Canonical link: <a href="https://commits.webkit.org/272969@main">https://commits.webkit.org/272969@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/82cc3e4815dac7f994a538e8654fbe705731223d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33752 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12529 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35695 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36372 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30605 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34804 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14904 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9680 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/29689 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34233 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10574 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30100 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9231 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/9333 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30117 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37694 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30643 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30438 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/35452 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9462 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7398 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33342 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11240 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/29776 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7796 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10042 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10247 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/avfoundation/WebAVSampleBufferListener.h:
```diff
@@ -0,0 +1,63 @@
+    void beginObservingRenderer(AVSampleBufferAudioRenderer*);
+    void stopObservingRenderer(AVSampleBufferAudioRenderer*);
+private:
+    WebAVSampleBufferListener(WebAVSampleBufferListenerClient&);
```
explicit

youennf commented on Source/WebCore/platform/graphics/avfoundation/WebAVSampleBufferListener.mm:
```diff
@@ -0,0 +1,295 @@
+    if (!(self = [super init]))
+        return nil;
+
+    _client = &client;
```
```suggestion
    _client = client;
```

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
```diff
@@ -124,19 +125,18 @@ class SourceBufferPrivateAVFObjC final
+    void outputObscuredDueToInsufficientExternalProtectionChanged(bool) final;
+    void layerRequiresFlushToResumeDecodingChanged(AVSampleBufferDisplayLayer *, bool) final;
+    void layerReadyForDisplayChanged(AVSampleBufferDisplayLayer *, bool isReadyForDisplay) final;
+    void rendererDidReceiveError(AVSampleBufferAudioRenderer *, NSError *) final;
```
Could probably be made private.

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
```diff
@@ -256,6 +262,8 @@ class MediaPlayerPrivateWebM
     static void getSupportedTypes(HashSet<String>&);
     static MediaPlayer::SupportsType supportsType(const MediaEngineSupportParameters&);
 
+    WeakPtr<WebMResourceClientParent> getWeakPtr() const { return *this; }
```
Why do we need this?
Can we instead use `using WebMResourceClientParent::weakPtrFactory` et al instead?

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -1551,6 +1561,21 @@ void getSupportedTypes(HashSet<String>& types) const final
+
+void MediaPlayerPrivateWebM::layerDidReceiveError(AVSampleBufferDisplayLayer*, NSError*)
+{
+    setNetworkState(MediaPlayer::NetworkState::DecodeError);
```
We could protect-this ourselves since the caller is not doing it, ditto below and for SourceBufferPrivate.

jyavenard replied:
changing the networkstate will not cause the parent to delete us.

youennf commented on Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
```diff
@@ -404,6 +404,7 @@
 
     --m_framesBeingDecoded;
     m_isDecodingSample = false;
+    m_lastDecodingError = noErr;
```
Why is it needed?

jyavenard replied:
this is to prevent the error callback to be called over and over for all pending decoding, even if the following decodes are successful. Will amend the commit log

youennf commented on LayoutTests/media/media-vp8-webm-error.html:
```diff
@@ -0,0 +1,36 @@
+</body>
+</html>
+
+
```
We can remove those lines.
This could also be a valid WPT test, with a few changes.

youennf commented on Source/WebCore/platform/graphics/avfoundation/WebAVSampleBufferListener.mm:
```diff
@@ -0,0 +1,295 @@
+        if (!_layers.contains(layer.get()))
+            return;
+        if (auto client = _client.get())
+            client->layerReadyForDisplayChanged(layer.get(), isReadyForDisplay);
```
I still think it would be good to protect client as nobody is ensuring client will remain valid during the whole call.
Maybe instead of protecting this inside the call, we should protect here:
`if (RefPtr client = _client.get())`.

We would have to add ref()/deref() virtual methods in the client interface.

jyavenard replied:
For now this is the same as the original code that was moved from SourceBufferPrivateAVFObjC and the API is designed to be main thread only, just as how SourceBufferPrivateAVFObjC currently is.

youennf commented on Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
```diff
@@ -404,6 +404,7 @@
 
     --m_framesBeingDecoded;
     m_isDecodingSample = false;
+    m_lastDecodingError = noErr;
```
Could you describe this in the commit message?

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -814,25 +817,25 @@ static bool isCopyDisplayedPixelBufferAvailable()
     
     if (auto itAudioRenderer = m_audioRenderers.find(trackId); itAudioRenderer != m_audioRenderers.end()) {
-        WeakPtr weakThis { *this };
+        WeakPtr weakThis { *static_cast<WebMResourceClientParent*>(this) };
```
Needed?

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -1358,13 +1365,14 @@ static bool isCopyDisplayedPixelBufferAvailable()
     characteristicsChanged();
 
-    WeakPtr weakThis { *this };
+    WeakPtr weakThis { *static_cast<WebMResourceClientParent*>(this) };
```
Needed?

jyavenard committed:
MediaPlayerPrivateWebM doesn't handle decoding errors.
https://bugs.webkit.org/show_bug.cgi?id=267165
rdar://120570295

Reviewed by Youenn Fablet.

The MediaPlayerPrivateWebM didn't listen for error from either the AVSampleBufferDisplayLayer
nor the AVSampleBufferAudioRenderer.
We extract the code from the SourceBufferPrivateAVFObjC and create a new WebAVSampleBufferListener
class.

Add test.

* LayoutTests/media/content/garbage.webm: Added.
* LayoutTests/media/media-vp8-webm-error-expected.txt: Added.
* LayoutTests/media/media-vp8-webm-error.html: Added.
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/avfoundation/WebAVSampleBufferListener.h: Added.
(WebCore::WebAVSampleBufferListenerClient::layerDidReceiveError):
(WebCore::WebAVSampleBufferListenerClient::layerRequiresFlushToResumeDecodingChanged):
(WebCore::WebAVSampleBufferListenerClient::layerReadyForDisplayChanged):
(WebCore::WebAVSampleBufferListenerClient::rendererDidReceiveError):
(WebCore::WebAVSampleBufferListenerClient::rendererWasAutomaticallyFlushed):
(WebCore::WebAVSampleBufferListenerClient::outputObscuredDueToInsufficientExternalProtectionChanged):
* Source/WebCore/platform/graphics/avfoundation/WebAVSampleBufferListener.mm: Added.
(-[WebAVSampleBufferListenerPrivate initWithClient:]):
(-[WebAVSampleBufferListenerPrivate invalidate]):
(-[WebAVSampleBufferListenerPrivate beginObservingLayer:]):
(-[WebAVSampleBufferListenerPrivate stopObservingLayer:]):
(-[WebAVSampleBufferListenerPrivate beginObservingRenderer:]):
(-[WebAVSampleBufferListenerPrivate stopObservingRenderer:]):
(-[WebAVSampleBufferListenerPrivate observeValueForKeyPath:ofObject:change:context:]):
(-[WebAVSampleBufferListenerPrivate layerFailedToDecode:]):
(-[WebAVSampleBufferListenerPrivate layerRequiresFlushToResumeDecodingChanged:]):
(-[WebAVSampleBufferListenerPrivate layerReadyForDisplayChanged:]):
(-[WebAVSampleBufferListenerPrivate audioRendererWasAutomaticallyFlushed:]):
(WebCore::WebAVSampleBufferListener::WebAVSampleBufferListener):
(WebCore::WebAVSampleBufferListener::invalidate):
(WebCore::WebAVSampleBufferListener::beginObservingLayer):
(WebCore::WebAVSampleBufferListener::stopObservingLayer):
(WebCore::WebAVSampleBufferListener::beginObservingRenderer):
(WebCore::WebAVSampleBufferListener::stopObservingRenderer):
* Source/WebCore/platform/graphics/avfoundation/objc/AudioTrackPrivateMediaSourceAVFObjC.cpp:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::SourceBufferPrivateAVFObjC):
(WebCore::SourceBufferPrivateAVFObjC::destroyRenderers):
(WebCore::SourceBufferPrivateAVFObjC::trackDidChangeEnabled):
(WebCore::SourceBufferPrivateAVFObjC::setVideoLayer):
(-[WebAVSampleBufferListener initWithParent:]): Deleted.
(-[WebAVSampleBufferListener dealloc]): Deleted.
(-[WebAVSampleBufferListener invalidate]): Deleted.
(-[WebAVSampleBufferListener beginObservingLayer:]): Deleted.
(-[WebAVSampleBufferListener stopObservingLayer:]): Deleted.
(-[WebAVSampleBufferListener beginObservingRenderer:]): Deleted.
(-[WebAVSampleBufferListener stopObservingRenderer:]): Deleted.
(-[WebAVSampleBufferListener observeValueForKeyPath:ofObject:change:context:]): Deleted.
(-[WebAVSampleBufferListener layerFailedToDecode:]): Deleted.
(-[WebAVSampleBufferListener layerRequiresFlushToResumeDecodingChanged:]): Deleted.
(-[WebAVSampleBufferListener layerReadyForDisplayChanged:]): Deleted.
(-[WebAVSampleBufferListener audioRendererWasAutomaticallyFlushed:]): Deleted.
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
(WebCore::MediaPlayerPrivateWebM::getWeakPtr const):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::MediaPlayerPrivateWebM):
(WebCore::MediaPlayerPrivateWebM::~MediaPlayerPrivateWebM):
(WebCore::MediaPlayerPrivateWebM::setDuration):
(WebCore::MediaPlayerPrivateWebM::notifyClientWhenReadyForMoreSamples):
(WebCore::MediaPlayerPrivateWebM::provideMediaData):
(WebCore::MediaPlayerPrivateWebM::trackDidChangeSelected):
(WebCore::MediaPlayerPrivateWebM::didParseInitializationData):
(WebCore::MediaPlayerPrivateWebM::append):
(WebCore::MediaPlayerPrivateWebM::ensureLayer):
(WebCore::MediaPlayerPrivateWebM::ensureDecompressionSession):
(WebCore::MediaPlayerPrivateWebM::addAudioRenderer):
(WebCore::MediaPlayerPrivateWebM::destroyLayer):
(WebCore::MediaPlayerPrivateWebM::destroyAudioRenderer):
(WebCore::MediaPlayerPrivateWebM::registerNotifyWhenHasAvailableVideoFrame):
(WebCore::MediaPlayerPrivateWebM::startVideoFrameMetadataGathering):
(WebCore::MediaPlayerPrivateWebM::layerDidReceiveError):
(WebCore::MediaPlayerPrivateWebM::rendererDidReceiveError):
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::finishCurrentDecodingAndReportError):

should we continue to decode data, a new error won't be produced.

Canonical link: https://commits.webkit.org/272969@main
